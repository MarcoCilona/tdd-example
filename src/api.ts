import axios from 'axios';
import httpAdapter from 'axios/lib/adapters/http';

const instance = axios.create({
  baseURL: 'https://api.github.com',
  adapter: process.env.NODE_ENV === 'test' && httpAdapter
})

export default {
  searchUser(username: string) {
    return instance
      .get(`/users/${username}`)
      .then(result => result.data)
  }
}
