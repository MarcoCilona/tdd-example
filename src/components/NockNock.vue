<template>
<div>
  <p data-test-id='typeParagraph'>{{ user }}</p>
  <button data-test-id='userButton' @click='handleClick'>Retrieve user!</button>
</div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';
import api from '@/api.ts';

@Component
export default class NockNock extends Vue {
  user = '';

  async handleClick() {
    await api.searchUser('Castrogiovanni');
    this.user = 'Digitiamo';
  }
}
</script>
