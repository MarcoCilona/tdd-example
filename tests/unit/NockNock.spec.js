import { done, shallowMount } from '@vue/test-utils';
import flushPromises from 'flush-promises';
import nock from 'nock';
import api from '@/api.ts';
import NockNock from '@/components/NockNock.vue';
import userFixture from './fixtures/User';

// Components

describe('Testing NockNock component', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(NockNock, {
      data: () => ({
        user: '',
      }),
    });
  });

  it('NockNock component is consistent', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Check get user api', async () => {
    const expectedUser = 'Digitiamo';

    const request = nock('https://api.github.com')
      .get(`/users/${expectedUser}`)
      .reply(200, userFixture);

    const result = await api.searchUser(expectedUser);
    await flushPromises();

    expect(result.name).toEqual(userFixture.name);
    expect(request.isDone()).toBe(true);
  });

  it('Check paragraph for user type exists', () => {
    expect(wrapper.find('[data-test-id="typeParagraph"]').exists()).toBe(true);
  });

  it('Check NockNock component has user data', () => {
    expect(wrapper.vm.$data.user).not.toBeUndefined();
  });

  it('Check button for retrieving user exists', () => {
    expect(wrapper.find('[data-test-id="userButton"]').exists()).toBe(true);
  });

  it('Check click on button updates the user object', async () => {
    const expectedUser = 'Castrogiovanni';
    nock('https://api.github.com')
      .get(`/users/${expectedUser}`)
      .reply(200, userFixture);
    const button = wrapper.find('[data-test-id="userButton"]');
    await button.trigger('click');

    await flushPromises();
    setTimeout(() => {
      expect(wrapper.find('[data-test-id="typeParagraph"]').element.textContent).toEqual(userFixture.name);
    }, 200);
  });
});
