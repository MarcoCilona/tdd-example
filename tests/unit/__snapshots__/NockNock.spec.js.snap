// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing NockNock component NockNock component is consistent 1`] = `
<div>
  <p data-test-id="typeParagraph"></p> <button data-test-id="userButton">Retrieve user!</button>
</div>
`;
